### ISSUE TITLE
seqlevels() 

### ISSUE THREAD

**Issue by mniederhuber** [snystrom/didactic-potato#4]

**Comment by mniederhuber** [comment]
Playing around with `seqlevels()`...

do `seqlevels()` and `seqinfo` introduce errors when matching levels between objects before `rtracklayer::summary()`?

Two GRanges objects with the same sequences but with 1 different range
```
A <- GRanges(c('chr1','chr2','chr3'), c(1:2,4:5,29:30))
B <- GRanges(c('chr3','chr1','chr2'), c(5:6,1:2,4:5))
seqlevels(A)
```
>[1] "chr1" "chr2" "chr3"

Mapping `seqlevels` of B to A  doesn't change ranges of A...
```
seqlevels(A) <- seqlevels(B)
A
```
>GRanges object with 3 ranges and 0 metadata columns:
      seqnames    ranges strand
         <Rle> <IRanges>  <Rle>
  [1]     chr1       1-2      *
  [2]     chr2       4-5      *
  [3]     chr3     29-30      *
>seqinfo: 3 sequences from an unspecified genome; no seqlengths

But does change levels order
```
seqlevels(A)
```
>[1] "chr3" "chr1" "chr2"

If there are sequences that are used in A but not present in B...
```
A <- GRanges(c('chr1','chr2','chr3','chrX'), c(1:2,4:5,29:30,0:1))
B <- GRanges(c('chr3','chr1','chr2'), c(5:10,1:2,4:5))

seqlevels(A) <- seqlevels(B)
```
>Error in GenomeInfoDb:::getDanglingSeqlevels(x, new2old = new2old, pruning.mode = pruning.mode, : The following seqlevels are to be dropped but are currently in use (i.e. have ranges on them): chrX. Please use the 'pruning.mode' argument to control how to prune 'x', that is, how to remove the ranges in 'x' that are on these sequences. For example, do something like: seqlevels(x, pruning.mode="coarse") <- new_seqlevels or: keepSeqlevels(x, new_seqlevels, pruning.mode="coarse") See ?seqinfo for a description of the pruning modes.

In reciprocal case...
```
seqlevels(B) <- seqlevels(A)
seqlevels(B)
```
>[1] "chr1" "chr2" "chr3" "chrX"

The 'chrX' sequence is simply added to B, and has no impact on existing ranges. 
```
B
```
>GRanges object with 3 ranges and 0 metadata columns:
      seqnames    ranges strand
         <Rle> <IRanges>  <Rle>
  [1]     chr3      5-10      *
  [2]     chr1       1-2      *
  [3]     chr2       4-5      *
>seqinfo: 4 sequences from an unspecified genome; no seqlengths

This may or may not be useful, but it doesn't seem like `seqlevels(A) <- seqlevels(B)` will introduce errors by manipulating existing ranges, and as long as we keep `pruning.mode = "error"` (I think that's default) - we should avoid inadvertently dropping ranges -- I need to test what an error here looks like within  `signal::get_bw_matrix()`
    - May be a good idea to add some kind of error check - `tryCatch()`?

**Comment by mniederhuber** [comment]
`signal::get_bw_matrix()` will produce the same `pruning.mode` error in case where `regions` has sequences not present in `bw` -- this will also happen if `bw` and `regions` have the same sequences but different sequence styles 
eg. `chr2L` vs `2L`

add something to remap all levels to same style?
```
newStyle <- mapSeqlevels(seqlevels(A), "NCBI")
A <- renameSeqlevels(A, newStyle)
```

**Comment by snystrom** [comment]
Woah, awesome! This is great, thanks for running this down! We should think about unit tests we should write to ensure this behavior does not change in the future. 

I like the idea of remapping to the same namespace. Although I wonder if doing it automatically could cause silent errors, especially for non-standard genomes. I think not. The way this should work is:

1. Save the original seqlevels of regions
2. Map regions and bw to NIH coordinates (or, can we map regions to bw coords?)
3. Remap the original seqlevels if needed (might be superfluous)

**Comment by mniederhuber** [comment]
Ya I think this will primarily be a problem when pulling existing datasets that aren't user generated -- more likely seqname styles will differ 

Probably simpler (for us) to map everything to the same standard style rather than detect existing style and map to that
what do you think about making this an optional argument -- to avoid forcing a remapping in cases where there may not be a matching style
```
get_bw_matrix(bw, regions, type = c("mean", "min", "max", "coverage", "sd"), remap = c(NULL,"NCBI","UCSC","Ensembl")
```

**Comment by 14stutzmanav** [comment]
I'm not sure if this is related to your ongoing conversation about seqlevels, but tonight I had an error that I was able to troubleshoot so I figured I'd share here. Basically, I kept getting this error:
```
get_average_signal(import.bw(paths.to.bigwigs[[5]]), plus.strand.data.peri)
Error in UseMethod("get_bw_matrix") : 
  no applicable method for 'get_bw_matrix' applied to an object of class "c('GRanges', 'GenomicRanges', 'GRanges_OR_NULL', 'Ranges', 'GenomicRanges_OR_missing', 'GenomicRanges_OR_GenomicRangesList', 'GenomicRanges_OR_GRangesList', 'List', 'Vector', 'list_OR_List', 'Annotated', 'vector_OR_Vector')"
```

So after some playing around with my version of didactic-potato, I found that this was caused because my regions and bw inputs did not match. To fix it, I edited the part called "#match seqlevels between inputs". It looked like this previously:
```
#match seqlevels between inputs 
  seqlevels(regions) <- seqlevels(bw)
  seqinfo(regions") <- seqinfo(bw)
```

Adding pruning.mode='coarse' worked for me. I realize this means some of my bigwig file will be lost, but it's stuff I don't care about because the lost data is on chromosomes that aren't in the regions file anyways. So I figured this was a viable solution. Here's what I changed the section to:
```
  seqlevels(regions, pruning.mode="coarse") <- seqlevels(bw)
  seqinfo(regions, pruning.mode="coarse") <- seqinfo(bw)
```

**Comment by mniederhuber** [comment]
first off, I love that you're using this. we should finish this package...

But heads up:
A quick glance at https://web.mit.edu/~r/current/arch/i386_linux26/lib/R/library/GenomeInfoDb/html/seqinfo.html 
makes me concerned this may not be doing exactly what you think. 

My read is that it's the `regions` that will get pruned, not the `bw`. 
Also, the description of `coarse` makes it sound like (depending on the format of the regions input, you may not just drop regions on missing chroms but a whole granges object. 
ie. if regions is a grangeslist object, where each object is a different set of peaks, then I think the whole object would get dropped from the list. 

```
"coarse": Remove the elements in x where the seqlevels to drop are in use. 
Typically reduces the length of x. Note that if x is a list-like object 
... any list element in x where at least one of the sequence levels to drop is in use is fully removed.
In other words, when pruning.mode="coarse", 
the seqlevels setter will keep or remove full list elements and not try to change their content. 
```
It may not be a problem if `regions` is a single granges object.

Solution might be to simply add a check to determine what type `regions` is before doing the correction.

**Comment by snystrom** [comment]
We totally should finish this thing, haha.

I can poke around tonight, too.
